### DISCUSSION TITLE
TimerQueue中两个set问题?

### DISCUSSION THREAD

**Discussion by cd106224** [chenshuo/muduo#566] [General]
Muduo库在TimerQueue中为什么要设置两个set,我发现 ActiveTimerSet是按地址排序的,注释上写时为了cancel,但是在cancel函数中也可以用TimerList来find进行取消,请问用两个set的初衷是什么
![image](https://user-images.githubusercontent.com/27359980/163947519-36b5f7b2-ff54-4d29-8216-94c6e5554b0b.png)

**Comment by chenshuo** [comment]
TimerQueue::cancelInLoop() 中 ActiveTimerSet::find() 的时间复杂度是 O(log N)；
如果改用 TimerList::find()，那么时间复杂度会增加到 O(N)，因为你不知道 Timerstamp。
你也不能通过 Timer* 找到 Timestamp，这里的 Timer* 不一定指向合法的对象，因为这个 Timer 对象可能已经析构了。
